File: client/go/BrowserLauncher.go
package modcdp

import (
	"fmt"
	"os"
	"strings"
)

type LaunchedBrowser struct {
	CDPURL                string   `json:"cdp_url,omitempty"`
	WSURL                 string   `json:"ws_url,omitempty"`
	Close                 func()   `json:"-"`
	ProfileDir            string   `json:"profile_dir,omitempty"`
	PipeRead              *os.File `json:"-"`
	PipeWrite             *os.File `json:"-"`
	BrowserbaseSessionID  string   `json:"browserbase_session_id,omitempty"`
	BrowserbaseSessionURL string   `json:"browserbase_session_url,omitempty"`
	BrowserbaseDebugURL   string   `json:"browserbase_debug_url,omitempty"`
}

type BrowserLauncher struct {
	Options  LaunchOptions
	Launched *LaunchedBrowser
}

func NewBrowserLauncher(options LaunchOptions) BrowserLauncher {
	return BrowserLauncher{Options: options}
}

func (l *BrowserLauncher) Update(config LaunchOptions) *BrowserLauncher {
	l.Options = mergeLaunchOptions(l.Options, config)
	return l
}

func (l BrowserLauncher) GetTransportConfig() map[string]any {
	return map[string]any{
		"cdp_url":       firstString(launchedCDPURL(l.Launched), l.Options.CDPURL),
		"ws_url":        firstString(launchedWSURL(l.Launched), l.Options.WSURL),
		"user_data_dir": firstString(launchedProfileDir(l.Launched), l.Options.UserDataDir),
		"pipe_read":     launchedPipeRead(l.Launched),
		"pipe_write":    launchedPipeWrite(l.Launched),
	}
}

func (l BrowserLauncher) GetInjectorConfig() ExtensionInjectorConfig {
	return ExtensionInjectorConfig{
		BrowserbaseAPIKey:  l.Options.BrowserbaseAPIKey,
		BaseURL:            l.Options.BaseURL,
		BrowserbaseBaseURL: l.Options.BrowserbaseBaseURL,
		ExtensionID:        l.Options.ExtensionID,
	}
}

func (l BrowserLauncher) Launch(options LaunchOptions) (*LaunchedBrowser, error) {
	return nil, fmt.Errorf("%T.Launch is not implemented", l)
}

func mergeLaunchOptions(existing LaunchOptions, incoming LaunchOptions) LaunchOptions {
	merged := existing
	if incoming.ExecutablePath != "" {
		merged.ExecutablePath = incoming.ExecutablePath
	}
	if incoming.Port != 0 {
		merged.Port = incoming.Port
	}
	if incoming.RemoteDebugging != "" {
		merged.RemoteDebugging = incoming.RemoteDebugging
	}
	if incoming.UserDataDir != "" {
		merged.UserDataDir = incoming.UserDataDir
	}
	if incoming.CleanupUserDataDir != nil {
		merged.CleanupUserDataDir = incoming.CleanupUserDataDir
	}
	if incoming.ChromeReadyTimeoutMS != 0 {
		merged.ChromeReadyTimeoutMS = incoming.ChromeReadyTimeoutMS
	}
	if incoming.ChromeReadyPollIntervalMS != 0 {
		merged.ChromeReadyPollIntervalMS = incoming.ChromeReadyPollIntervalMS
	}
	if incoming.Headless != nil {
		merged.Headless = incoming.Headless
	}
	if incoming.Sandbox != nil {
		merged.Sandbox = incoming.Sandbox
	}
	if len(incoming.Args) > 0 {
		merged.Args = mergeChromeArgs(existing.Args, incoming.Args)
	}
	if len(incoming.ExtraArgs) > 0 {
		merged.ExtraArgs = mergeChromeArgs(existing.ExtraArgs, incoming.ExtraArgs)
	}
	if incoming.CDPURL != "" {
		merged.CDPURL = incoming.CDPURL
	}
	if incoming.WSURL != "" {
		merged.WSURL = incoming.WSURL
	}
	if incoming.BrowserbaseAPIKey != "" {
		merged.BrowserbaseAPIKey = incoming.BrowserbaseAPIKey
	}
	if incoming.ProjectID != "" {
		merged.ProjectID = incoming.ProjectID
	}
	if incoming.BrowserbaseProjectID != "" {
		merged.BrowserbaseProjectID = incoming.BrowserbaseProjectID
	}
	if incoming.BaseURL != "" {
		merged.BaseURL = incoming.BaseURL
	}
	if incoming.BrowserbaseBaseURL != "" {
		merged.BrowserbaseBaseURL = incoming.BrowserbaseBaseURL
	}
	if incoming.SessionID != "" {
		merged.SessionID = incoming.SessionID
	}
	if incoming.ResumeSessionID != "" {
		merged.ResumeSessionID = incoming.ResumeSessionID
	}
	if incoming.KeepAlive != nil {
		merged.KeepAlive = incoming.KeepAlive
	}
	if incoming.CloseSessionOnClose != nil {
		merged.CloseSessionOnClose = incoming.CloseSessionOnClose
	}
	if incoming.Region != "" {
		merged.Region = incoming.Region
	}
	if incoming.Timeout != 0 {
		merged.Timeout = incoming.Timeout
	}
	if incoming.ExtensionID != "" {
		merged.ExtensionID = incoming.ExtensionID
	}
	if incoming.BrowserSettings != nil {
		merged.BrowserSettings = incoming.BrowserSettings
	}
	if incoming.UserMetadata != nil {
		merged.UserMetadata = incoming.UserMetadata
	}
	if incoming.SessionCreateParams != nil {
		merged.SessionCreateParams = incoming.SessionCreateParams
	}
	if incoming.BrowserbaseSessionCreateParams != nil {
		merged.BrowserbaseSessionCreateParams = incoming.BrowserbaseSessionCreateParams
	}
	return merged
}

func mergeChromeArgs(existing []string, incoming []string) []string {
	args := append(append([]string{}, existing...), incoming...)
	loadExtensionPaths := []string{}
	merged := []string{}
	for _, arg := range args {
		if !strings.HasPrefix(arg, "--load-extension=") {
			merged = append(merged, arg)
			continue
		}
		for _, extensionPath := range strings.Split(strings.TrimPrefix(arg, "--load-extension="), ",") {
			if extensionPath != "" && !containsString(loadExtensionPaths, extensionPath) {
				loadExtensionPaths = append(loadExtensionPaths, extensionPath)
			}
		}
	}
	if len(loadExtensionPaths) > 0 {
		loadExtensionArg := "--load-extension=" + strings.Join(loadExtensionPaths, ",")
		firstURLIndex := -1
		for index, arg := range merged {
			if !strings.HasPrefix(arg, "-") {
				firstURLIndex = index
				break
			}
		}
		if firstURLIndex == -1 {
			merged = append(merged, loadExtensionArg)
		} else {
			merged = append(merged[:firstURLIndex], append([]string{loadExtensionArg}, merged[firstURLIndex:]...)...)
		}
	}
	return merged
}

func containsString(values []string, needle string) bool {
	for _, value := range values {
		if value == needle {
			return true
		}
	}
	return false
}

func firstString(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

func launchedCDPURL(launched *LaunchedBrowser) string {
	if launched == nil {
		return ""
	}
	return launched.CDPURL
}

func launchedWSURL(launched *LaunchedBrowser) string {
	if launched == nil {
		return ""
	}
	return launched.WSURL
}

func launchedProfileDir(launched *LaunchedBrowser) string {
	if launched == nil {
		return ""
	}
	return launched.ProfileDir
}

func launchedPipeRead(launched *LaunchedBrowser) *os.File {
	if launched == nil {
		return nil
	}
	return launched.PipeRead
}

func launchedPipeWrite(launched *LaunchedBrowser) *os.File {
	if launched == nil {
		return nil
	}
	return launched.PipeWrite
}

File: client/python/modcdp/BrowserLauncher.py
from __future__ import annotations

from typing import Any, TypedDict, cast


class BrowserLaunchOptions(TypedDict, total=False):
    executable_path: str | None
    port: int | None
    user_data_dir: str | None
    headless: bool
    sandbox: bool
    args: list[str]
    extra_args: list[str]
    remote_debugging: str
    cleanup_user_data_dir: bool
    chrome_ready_timeout_ms: int
    chrome_ready_poll_interval_ms: int
    cdp_url: str | None
    ws_url: str | None
    browserbase_api_key: str | None
    project_id: str | None
    browserbase_project_id: str | None
    base_url: str | None
    browserbase_base_url: str | None
    session_id: str | None
    resume_session_id: str | None
    keep_alive: bool
    close_session_on_close: bool
    region: str | None
    timeout: int | None
    extension_id: str | None
    browser_settings: dict[str, Any] | None
    user_metadata: dict[str, Any] | None
    session_create_params: dict[str, Any] | None
    browserbase_session_create_params: dict[str, Any] | None


class LaunchedBrowser(TypedDict, total=False):
    cdp_url: str | None
    ws_url: str | None
    close: Any
    profile_dir: str | None
    pipe_read: Any
    pipe_write: Any


DEFAULT_CHROME_READY_TIMEOUT_MS = 45_000
DEFAULT_CHROME_READY_POLL_INTERVAL_MS = 100


class BrowserLauncher:
    launched: LaunchedBrowser | None

    def __init__(self, options: BrowserLaunchOptions | None = None) -> None:
        self.options = cast(BrowserLaunchOptions, dict(options or {}))
        self.launched = None

    def update(self, config: BrowserLaunchOptions | None = None) -> "BrowserLauncher":
        config = cast(BrowserLaunchOptions, dict(config or {}))
        self.options = cast(
            BrowserLaunchOptions,
            {
                **self.options,
                **config,
                **({"args": merge_chrome_args(self.options.get("args"), config["args"])} if "args" in config else {}),
                **(
                    {"extra_args": merge_chrome_args(self.options.get("extra_args"), config["extra_args"])}
                    if "extra_args" in config
                    else {}
                ),
            },
        )
        return self

    def getTransportConfig(self) -> dict[str, Any]:
        return {
            "cdp_url": (self.launched or {}).get("cdp_url") or self.options.get("cdp_url"),
            "ws_url": (self.launched or {}).get("ws_url") or self.options.get("ws_url"),
            "user_data_dir": (self.launched or {}).get("profile_dir") or self.options.get("user_data_dir"),
            "pipe_read": (self.launched or {}).get("pipe_read"),
            "pipe_write": (self.launched or {}).get("pipe_write"),
        }

    def getInjectorConfig(self) -> dict[str, Any]:
        return {
            "browserbase_api_key": self.options.get("browserbase_api_key"),
            "base_url": self.options.get("base_url"),
            "browserbase_base_url": self.options.get("browserbase_base_url"),
            "extension_id": self.options.get("extension_id"),
        }

    def launch(self, options: BrowserLaunchOptions | None = None) -> LaunchedBrowser:
        raise NotImplementedError(f"{type(self).__name__}.launch is not implemented.")


def merge_chrome_args(existing: list[str] | None = None, incoming: list[str] | None = None) -> list[str]:
    args = [*(existing or []), *(incoming or [])]
    load_extension_paths: list[str] = []
    merged: list[str] = []
    for arg in args:
        if not arg.startswith("--load-extension="):
            merged.append(arg)
            continue
        for extension_path in arg[len("--load-extension="):].split(","):
            if extension_path and extension_path not in load_extension_paths:
                load_extension_paths.append(extension_path)
    if load_extension_paths:
        first_url_index = next((index for index, arg in enumerate(merged) if not arg.startswith("-")), -1)
        load_extension_arg = f"--load-extension={','.join(load_extension_paths)}"
        if first_url_index == -1:
            merged.append(load_extension_arg)
        else:
            merged.insert(first_url_index, load_extension_arg)
    return merged
